src/common/args.h

```diff
@@ -152,12 +224,73 @@ class ArgsManager
     bool UseDefaultSection(const std::string& arg) const EXCLUSIVE_LOCKS_REQUIRED(cs_args);
 
  public:
+    ArgsManager();
+    ~ArgsManager();
+
+    /**
+     * @name GetArg Functions
+     *
+     * GetArg functions are an easy way to access most settings. They are
+     * wrappers around the lower-level GetSetting() function that provide
+     * greater convenience.
+     *
+     * Examples:
+     *
+     *     GetArg("-foo")     // returns "abc" if -foo=abc was specified, or nullopt if unset
+     *     GetIntArg("-foo")  // returns 123 if -foo=123 was specified, or nullopt if unset
+     *     GetBoolArg("-foo") // returns true if -foo was specified, or nullopt if unset
+     *     GetBoolArg("-foo") // returns false if -nofoo was specified, or nullopt if unset
+     *
+     * If no type flags (ALLOW_STRING, ALLOW_INT, or ALLOW_BOOL) are set, GetArg
+     * functions do many type coercions and can have surprising behaviors which
+     * legacy code relies on, like parsing -nofoo as string "0" or -foo=true as
+     * boolean false.
+     *
+     * If any type flags are set, then:
+     *
+     * - Only GetArg functions with types matching the flags can be called. For
+     *   example, it is an error to call GetIntArg() if ALLOW_INT is not set.
+     *
+     * - GetArg functions act like std::get_if<T>(), returning null if the
+     *   requested type is not available or the setting is unspecified or empty.
+     *
+     * - "Widening" type conversions from smaller to bigger types are done if
+     *   unambiguous (bool -> int -> string). For example, if settings.json
+     *   contains {"foo":123}, GetArg("-foo") will return "123". If it contains
+     *   {"foo":true}, GetIntArg("-foo") will return 1.
+     *
+     * - "Narrowing" type conversions in the other direction are not done even
+     *   when they would be unambiguous. This makes it possible to distinguish
+     *   values by type by checking for narrow types first. For example, to
+     *   handle boolean settings.json or command line values (-foo -nofoo)
+     *   differently than string values (-foo=abc), you can write:
+     *
+     *       if (auto foo{args.GetBoolArg("-foo")}) {
+     *           // handle -foo or -nofoo bool in foo.value()
+     *       } else if (auto foo{args.GetArg("-foo")}) {
+     *           // handle -foo=abc string in foo.value()
+     *       } else {
+     *           // handle unset setting
+     *       }
```

Comment:
Feels like this is encouraging `BOOL | STRING` usage and should be broken out too if we are postponing that support.